spark/src/main/java/org/apache/iceberg/actions/BaseAction.java

```diff
@@ -94,9 +102,30 @@ protected String metadataTableName(String tableName, MetadataTableType type) {
     return buildValidDataFileDF(spark, table().toString());
   }
 
+  private static class ReadManifest implements FlatMapFunction<ManifestFileBean, String> {
+    private final Broadcast<FileIO> io;
+
+    ReadManifest(Broadcast<FileIO> io) {
+      this.io = io;
+    }
+
+    @Override
+    public Iterator<String> call(ManifestFileBean manifest) {
+      return new ClosingIterator<>(ManifestFiles.readPaths(manifest, io.getValue()).iterator());
+    }
+  }
+
   protected Dataset<Row> buildValidDataFileDF(SparkSession spark, String tableName) {
-    String allDataFilesMetadataTable = metadataTableName(tableName, MetadataTableType.ALL_DATA_FILES);
-    return spark.read().format("iceberg").load(allDataFilesMetadataTable).select("file_path");
+    JavaSparkContext context = new JavaSparkContext(spark.sparkContext());
+    Broadcast<FileIO> ioBroadcast = context.broadcast(SparkUtil.serializableFileIO(table()));
```

Comment:
nit: We use `Broadcast<FileIO> io` in all other places.

Reply:
Sorry I missed fixing this. We can update it next time we touch this code.